File: LTNC/STL_Stack_Queue/TRUNGHAUTO2.cpp
#include <bits/stdc++.h>
using namespace std;

struct Stack {
    vector<string> a;
    void push(const string &s){ a.push_back(s); }
    string pop(){ string v=a.back(); a.pop_back(); return v; }
    string &top(){ return a.back(); }
    bool empty(){ return a.empty(); }
};

int prec(const string &op){
    if(op=="+"||op=="-") return 1;
    if(op=="*"||op=="/") return 2;
    if(op=="^") return 3;
    return 0;
}

int main(){
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    string line, all;
    while(true){
        string tmp;
        if(!getline(cin,tmp)) break;
        if(!all.empty()) all.push_back(' ');
        all += tmp;
    }
    if(all.empty()) return 0;
    for(char &c: all) if(c=='\u2013') c='-';
    stringstream ss(all);
    vector<string> tokens;
    string tk;
    while(ss>>tk) tokens.push_back(tk);

    Stack st;
    vector<string> out;
    for(auto &t: tokens){
        if(!t.empty() && all_of(t.begin(), t.end(), ::isdigit)){
            out.push_back(t);
        } else if(t=="("){
            st.push(t);
        } else if(t==")"){
            while(!st.empty() && st.top()!="("){
                out.push_back(st.pop());
            }
            if(!st.empty() && st.top()=="(") st.pop();
        } else { 
            while(!st.empty() && st.top()!="("){
                string &tt = st.top();
                int pt=prec(tt), pc=prec(t);
                if(pt>pc || (pt==pc && t!="^")){
                    out.push_back(st.pop());
                } else break;
            }
            st.push(t);
        }
    }
    while(!st.empty()) out.push_back(st.pop());
    for(size_t i=0;i<out.size();++i){
        if(i) cout<<' ';
        cout<<out[i];
    }
    return 0;
}

File: LTNC/STL_Stack_Queue/TRUNGHAUTO2.py
'''
Nhập vào 1 biêu thức trung tố, yêu cầu xuất biểu thức theo dạng hậu tố.

Input:

+ Nhập biểu thức trung tố

(lưu ý:

+ Các số hạng, phép toán cách nhau 1 khoảng trắng

+ mỗi số hạng <= 106)

Ouput:

  + Biểu thức trung tố (số hạng và phép toán cách nhau 1 khoảng trắng)



Ví dụ:

Input	
3 + 4 * 2 / ( 1 – 5 )
Output
3 4 2 * 1 5 - / +

'''

import sys

class Stack:
    def __init__(self):
        self.a=[]
    def push(self,x): self.a.append(x)
    def pop(self): return self.a.pop()
    def top(self): return self.a[-1]
    def empty(self): return not self.a

line = sys.stdin.read().strip()
if not line:
    sys.exit()

line = line.replace('–','-')
tokens = line.split()

prec = {'+':1,'-':1,'*':2,'/':2,'^':3}
right_assoc = {'^'}
st = Stack()
out = []

for tk in tokens:
    if tk.isdigit():               
        out.append(tk)
    elif tk == '(':
        st.push(tk)
    elif tk == ')':
        while not st.empty() and st.top()!='(':
            out.append(st.pop())
        if not st.empty() and st.top()=='(':
            st.pop()
    else:  
        while (not st.empty() and st.top()!='(' and
               (prec[st.top()]>prec[tk] or
                (prec[st.top()]==prec[tk] and tk not in right_assoc))):
            out.append(st.pop())
        st.push(tk)

while not st.empty():
    out.append(st.pop())

print(' '.join(out))